macos/Sources/App/macOS/AppDelegate.swift

```diff
@@ -317,14 +317,33 @@ class AppDelegate: NSObject,
         return derivedConfig.shouldQuitAfterLastWindowClosed
     }
 
+    /// Initiates graceful application termination.
+    private func terminateGracefully() -> NSApplication.TerminateReply {
+        /// Free our surfaces synchronously, ensuring SIGHUP signals are sent to all
+        /// child processes (e.g., bash) so they can clean up before the app exits.
+        TerminalController.freeAllSurfaces()
+
+        if !quickController.surfaceTree.isEmpty {
+            quickController.freeSurfaces()
+        }
+
+        // Schedule termination after a brief delay to allow child processes
+        // to handle SIGHUP and complete cleanup (e.g., bash saving history).
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+            NSApp.reply(toApplicationShouldTerminate: true)
+        }
+
+        return .terminateLater
```

Comment:
We could instead poll the child processes and wait for all of them to explicitly terminate, but that would require a bit more work across the libghostty boundary.

I think 100ms is a good, simple start, but we can explore that improvement separately if we want this to be an even more complete solution. I think we might also need to consider some user-visible "shutting down ..." UI feedback in that case.